### ISSUE TITLE
array() does not produce empty arrays with empty input

### ISSUE THREAD

**Issue by misje** [Velocidex/velociraptor#3606]
`array()` should produce `[]`, and not `[{}]`. The same goes for `array()` calls where the input results in an empty array. I'm pretty sure this worked before in older versions of velociraptor, otherwise I wouldn't have to update artifacts for this unexpected empty single-object array, which is no longer considered empty/false.

**Comment by scudette** [comment]
Can you give more examples about this case 
>  The same goes for array() calls where the input results in an empty array.

Array aims to flatten the input into a flat array. It does this by expanding each arg into a list, then flattening it in turn. This can be a little unexpected because in other languages the array() function is basically the same to the [] notation:

```
SELECT array(a=[dict(b=1), dict(c=2)], b=dict(d=5)) FROM scope()

[
   1,
   2,
   5
]

```

But the more obvious notation is 
```
SELECT ((dict(b=1), dict(c=2)), dict(d=5)) FROM scope()

  [
    {
     "b": 1
    },
    {
     "c": 2
    }
   ],
   {
    "d": 5
   }
  ]

```

Can you also elaborate on how you use the array() function? why is it better than just [] or () ?

Im not certain that the array() behavior is justified though it may be too complex and maybe should be renamed as "flatten" (which is a plugin that does similar work)

**Comment by misje** [comment]
If you call this method "flatten", than everything makes a lot more sense.

>  why is it better than just [] or () ?

I can't get that syntax to work. How do I create an empty array in the following example? `SELECT [] FROM scope()?`. The query I'm working at needs a condition, returning either a massive array from a JSON query, or if that query failed (looking at the response code), return an empty array. – In order to avoid errors later in the artifact. I figured `array()` would be a drop-in replacement when an empty array literal doesn't work ([], (), (,)?), but now that you describe its actual implementation/use, then what I truly need is just a way to express an empty array.

> Can you give more examples about this case

I can only think of one personally, which was [here](https://github.com/Velocidex/velociraptor-docs/blob/master/content/exchange/artifacts/Linux.Network.NM.Connections.yaml#L187). I wrapped that in filter instead.
